### PR TITLE
chore: introduce encoding factory

### PR DESCRIPTION
**Pull request by achettyiitr**
# Description

- Encoding manager: Deals with related tasks corresponding to encodings like reading, writing, and loading.
- Some code cleanup.
- Removed `encoding.Init()`.

## Linear Ticket

- https://linear.app/rudderstack/issue/PIPE-135/encoding-manager

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

**Review by fracasula — COMMENTED** [review]
Sorry for the partial review, I didn't have the time to complete it. Unfortunately I won't have the time to do so until Wednesday. In any case don't consider these comments as blockers, I just wanted to start a discussion.

File: warehouse/encoding/jsonreader.go [hunk 1]
```diff
@@ -7,46 +7,43 @@ import (
...
-		if err != nil {
+		if err := js.scanner.Err(); err != nil {
 			return []string{}, fmt.Errorf("scanner scan: %w", err)
 		}
 		return []string{}, io.EOF
 	}
 
 	lineBytes := js.scanner.Bytes()
-	jsonData := make(map[string]string)
+	jsonData := make(map[string]interface{})
```
Comment (achettyiitr) [inline warehouse/encoding/jsonreader.go:24]: This came across while writing the tests. It's expecting only string which will not be the case.

File: warehouse/encoding/encoding.go [hunk 2]
```diff
@@ -13,24 +18,69 @@ const (
...
+func (m *Manager) NewWriter(loadFileType, outputFilePath string, schema model.TableSchema, destType string) (Writer, error) {
+	switch loadFileType {
+	case warehouseutils.LoadFileTypeParquet:
+		return createParquetWriter(outputFilePath, schema, destType, m.config.parquetParallelWriters)
+	default:
+		return misc.CreateGZ(outputFilePath)
+	}
+}
+
+type Loader interface {
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/encoding/encoding.go:54]: Yeah, I completely aggress with having a solid interface.

File: warehouse/encoding/encoding.go [hunk 3]
```diff
@@ -13,24 +18,69 @@ const (
...
+	m := &Manager{}
+
+	conf.RegisterIntConfigVariable(10240, &m.config.maxStagingFileReadBufferCapacityInK, false, 1, "Warehouse.maxStagingFileReadBufferCapacityInK")
+	conf.RegisterInt64ConfigVariable(8, &m.config.parquetParallelWriters, true, 1, "Warehouse.parquetParallelWriters")
+
+	return m
+}
 
-type LoadFileWriter interface {
+type Writer interface {
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/encoding/encoding.go]: I was thinking in terms of the warehouse module only where:
1. Readers   -> Load file Readers.
2. Writers    -> Load file singular event writers.
3. Loaders  -> Load file singular event loaders.

**Review by lvrach — APPROVED** [review]

**Review by fracasula — APPROVED** [review]

**Pull request merged**